proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,641 @@
+used as Web IDL Types.
+
+As an example, the Web IDL Dictionary:
+
+```
+dictionary Contact {
+  DOMString name;
+  long age;
+};
+```
+
+could be defined in a WebAssembly module with this (strawman text-format)
+statement:
+
+```
+(webidl-type $Contact (dict (field "name" DOMString) (field "age" long)))
+```
+
+Following the general WebAssembly text format convention, `$Contact` may be
+used throughout the Web IDL Bindings Section, with uses converted to the
+integer index of the type definition in the Web IDL Types Subsection when the
+`.wat` is converted to `.wasm`.
+
+
+## Binding Operators and Expressions
+
+Web IDL values are manipulated via Web IDL **binding expressions** which, like
+core WebAssembly instructions, are defined in the form of [abstract syntax] with
+both text and binary formats that are composed of **binding operators** which
+can have immediate operands and child expressions.  Binding operators come in
+two flavors—**outgoing** and **incoming**—which correspond to the two
+directions that values flow through a binding (from a WebAssembly perspective):
+outgoing (to Web IDL) or incoming (from Web IDL).  The "outgoing" and
+"incoming" distinction is orthogonal to the "import" and "export" distinction:
+both import and export bindings use both outgoing and incoming binding
+expressions.
+
+An **outgoing binding expression** is composed of **outgoing binding
+operators** and maps from a **source tuple** of WebAssembly values to a
+**destination tuple** of Web IDL values (although currently only one is ever
+produced). A sample list of outgoing binding operators:
+
+| Operator | Immediates | Children | Description |
+| :---     | :---       | :---     | :---        |
+| **as** | webidl‑type<br>idx | | Takes the `idx`'th wasm value of the source tuple and produces a Web IDL value of `webidl-type`, allowing only trivial conversions (like `i32` to [`long`] or `anyref` to [`any`]). |
+| **utf8‑str** | webidl‑type<br>off‑idx<br>len‑idx | | Takes the `off-idx`'th and `len-idx`th wasm values of the source tuple, which must both be `i32`s, as the offset and length, resp., of a UTF-8 string in linear memory and decodes a `webidl-type`, which must be one of [`DOMString`], [`ByteString`] or [`USVString`]. |
+| **utf8‑cstr** | webidl‑type<br>off‑idx	| | Like `utf8-str`, but instead of length being explicitly supplied as a wasm value in the source tuple, the length is found by decoding up to the first null code point. |
+| **i32‑to‑enum** | webidl‑type<br>val‑idx	| | Takes the `val-idx`'th wasm value of the source tuple, which must be an `i32`, as an index into the list of strings specified by `webidl-type`, which must be an [Enumeration]. |
+| **view** | webidl‑type<br>off‑idx<br>len‑idx | | Takes the `off-idx`'th and `len-idx`'th wasm values of the source tuple, which must both be `i32`s, as the offset and length, resp., of a new `webidl-type`, which must be a [Typed Array View]. |
```

Comment:
I presume a `view` means that data is not copied out of linear memory, but referred to, whereas e.g. `utf8‑str` implies a copy. While not copying is the best default, is there a use case for a `buffer` type (meaning copied, but not utf8), for cases where it be very inconvenient for the wasm side to have to hold on to the data (because the browser side use is more long-lived, or even GC-ed) ?

Reply:
That's a great point and I actually had, in various previous revisions of this draft, a `buffer` operator that does exactly what you said.  I forgot why I cut it... anyway, it's back!